xarray/core/merge.py

```diff
@@ -365,6 +365,24 @@ def merge_data_and_coords(data, coords, compat='broadcast_equals',
     return merge_core(objs, compat, join, explicit_coords=explicit_coords)
 
 
+def assert_valid_explicit_coords(variables, explicit_coords):
+    """Validate explicit coordinate names/dims.
+
+    Raise a MergeError if an explicit coord shares a name with a dimension
+    but is comprised of arbitrary dimensions.
+    """
+    var_dims = []
```

Comment:
Use a set here.

Reply:
I don't follow. Are you just suggesting we cast `var_dims` to a set after populating it or are you looking for some different logic for determining the dimensions in the non-explicit coord variables?

Reply:
Just that a `set()` is a better data structure for contains lookups than a list.

But actually, you should use the result of `calculate_dimensions()` here, instead of calculating dimensions twice.